src/librustc/mir/visit.rs

```diff
@@ -792,26 +792,35 @@ macro_rules! visit_place_fns {
         ) {
             self.visit_place_base(&mut place.base, context, location);
 
-            place.projection = self.process_projection(&place.projection);
+            let cow_projection = self.process_projection(&place.projection);
+
+            // FIXME: using Cow doesn't make a lot of sense until we start interning projection.
+            place.projection = cow_projection.into_owned().into_boxed_slice();
```

Comment:
this might be more annoying but you could also use `Option` -- you return `Some` if there is a replacement, `None` if no changes. 

```rust
if let Some(replacement) = self.process_projection(&place.projection) {
    place.projection = replacement;
}

fn process_projection(&mut self, ...) -> Option<Box<[PlaceElem<'tcx>>> { ... }
```